### ISSUE TITLE
Add referenced links to Markdown

### ISSUE THREAD

**Issue by akwiatkowski** [crystal-lang/crystal#3656]
At this moment Crystal markdown parser cannot parse and use link references like this

```md
[link]: google.pl
[Search engine][link]
```

I have managed to create different way to render links, however there is a problem which I hadn't thought about. This example fails now
```
assert_render "Hello [world **2**](http://example.com)!", %(<p>Hello <a href="http://example.com">world <strong>2</strong></a>!</p>)
```

I wonder why Markdown parser is written by operating byte to byte. In my opinion it should search recursive. It could be not as fast or memory efficient, but it would be much easier to implement.

**Comment by asterite** [comment]
@akwiatkowski If you'd like to implement in a different way which supports all of markdown features and works well, and isn't terrible slow, please send a PR and we'll gladly review it and probably merge it :-)

This was my first markdown implementation. In the beginning I thought markdown was simple so I started with a simple  and efficient implementation. But it turned out it's very, very complex to parse and get right. I always try to do things in an efficient way. For example markdown parsing is used for generating docs, and this is currently very fast. If markdown parsing is slower, generating docs will be a bit slower. But if it's not terribly slower then I think it's fine.

**Comment by asterite** [comment]
Actually, with the current implementation this seems not-that-hard to implement. First we make a pass through all lines to gather links. Then as we traverse the text, if we find `[text]` se check if it's on the gathered links. Nothing else needs to change, right?

**Comment by akwiatkowski** [comment]
If I'm correct there can be:

1. Regular link `[WWW](http://example.com)`
2. Referenced without text: `[www]: http://example.com \n [WWW][www]`
3. Referenced with default text: `[www]: http://example.com WWW \n [www]`
4. Referenced with default text, but other is forced: `[www]: http://example.com WWW \n [Cool site][www]`

**Comment by ysbaddaden** [comment]
There is also this way for direct links:

```md
See more at <http://example.org>
```

**Comment by akwiatkowski** [comment]
I've changed way how it will be implemented. I assume that current version is at least ok and I don't want to change too much because links can be a little bit tricky (like nested with other tags). I've chosen "the preprocessor style". First it analyze all references and then it will overwrite referenced links to regular ones. Not all cases like images, and an example above are done yet, but they could be easily added if you accept this way.

Changes are at https://github.com/crystal-lang/crystal/pull/3686

**Comment by akwiatkowski** [comment]
There is problem with links with closing parenthesis at this moment. Fixed it locally by escaping but I'm not happy with this solution.

```crystal
escaped_url = url.gsub(/\)/, "%29")
```

Sample:

```markdown
[Link](https://pl.wikipedia.org/wiki/Por%C4%99ba_Wielka_(powiat_o%C5%9Bwi%C4%99cimski))
```

You can reproduce it

```crystal
require "markdown"

s = "[Link](https://pl.wikipedia.org/wiki/Por%C4%99ba_Wielka_(powiat_o%C5%9Bwi%C4%99cimski))"
puts Markdown.to_html(s)
```

**Comment by straight-shoota** [comment]
I think this could be closed in favour of #4613

**Comment by asterite** [comment]
See #4613
